### ISSUE TITLE
Make Tree Control Node More Usable

### ISSUE THREAD

**Issue by name-here** [godotengine/godot-proposals#2088]
### Describe the project you are working on

A plugin that adds a panel with a list of my game's levels, so I can easily re-order, rename, edit, and run them, and it will automatically update a custom Resource with any changes

### Describe the problem or limitation you are having in your project

Despite all of these (re-order, rename, etc.) already being features of the Scene dock, as far as I can tell, most of the features I need aren't exposed to GDScript.  Missing features I'd expect include:
- Getting the `TreeNode` at a particular index
- Adding an already-constructed `TreeNode` object to the `Tree`, instead of having to add a blank one with `create_item()` and then copy all the attributes one by one
- Relatedly, removing a TreeItem from the list without doing `.free()` (they don't have a `.duplicate()` function because they directly inherit `Object`)
- Re-ordering `TreeNode`s (moving from one index to another)
- Drag-and-drop functionality to let the user re-order `TreeNode`s (currently, you have to write your own functions for moving the nodes around, which is a pain because there's no function to re-order them through code, and the list isn't exposed for editing)

#### For re-ordering through script:
I can think of any number of ways to do this, some better than others.  Here are a few:
1. Expose the `next` and `previous` pointers on `TreeItem`s to GDScript, allowing the user to re-arrange the list to their liking (I think it's a linked list?).  This adds versatility, but IMO still requires too much work for the user, and leaves too much room for error in messing up the linked list.
2. Just expose functions for adding an already-constructed `TreeItem` at an index, moving a `TreeItem` from one index to another, and getting the `TreeItem` for a given index.  Also helpful would be a way to get the index of a given `TreeItem`.  This is probably the easiest "good" solution, but others might make `Tree` match better with other parts of the engine.
3. Make `TreeItem` inherit from `Node` (maybe even `Control`?), and have them actually be children of the `Tree`.  This is probably the most complex solution, and I'm not sure how the details would work out (`Control`s having their position set by a parent many levels up the `SceneTree`?), but it means `TreeItem` wouldn't have to duplicate all that functionality, and could be easily moved around just like other `Node`s.
4. Put all of a `Tree`'s `TreeItem`s in an `Array`, where each `TreeItem` has an `Array` for its children (if it has any).  I think this would be a good fix for the accessing and re-ordering problems, but without statically typed arrays, there's nothing stopping something of the wrong type from being inserted into the list.

#### For drag and drop:
It'd nice to be able to write your own custom drag-and-drop if you have special requirements for what can be dropped where and the like, but the fact that these functions (`get_drag_data()`, `can_drop_data()`, `drop_data()`) don't even have a default implementation is really frustrating.  I'd say there should be, at minimum, default implementations for all of those three, and a functional way to permanently set the drop mode flags from the editor.  It would also be nice to have a function that built a drag preview control for you from the dragged `TreeItem` (for `set_drag_preview()`).

#### Other things:
- `Tree` has an `item_edited()` signal…that doesn't tell you which item was edited.  And `get_edited()` and `get_edited_column()`, which seem like they might help, say "This is only available for custom cell mode" in the docs.  This at least shouldn't be a difficult fix, right?
  - Likewise with `item_rmb_edited()`
- This one's minor, but why doesn't the `item_selected()` signal just tell you which item it was?  Not a huge deal since `get_selected()` should work, but why not, when `multi_selected()` actually _does_ tell you?
  - The signals `cell_selected()`, `item_activated()`, `item_custom_button_pressed()`, and `item_double_clicked()`are like this too.  When would I want to know _that_ something was clicked, activated, etc., but not _what_?
- `Tree` has `are_column_titles_visible()` and `set_column_titles_visible()`.  These should be replaced with an exposed `colum_titles_visible` property with standard set and get functions.
- `TreeItem`  has a bunch of per-column properties that are exposed only through `set_x(column, data)` and `get_x(column)` functions.  To me, this feels kind of messy.  Ideas:
  - Have some datatype that holds all the data for a given column (maybe `TreeItemColumn` or something?), and give `TreeItem` an `Array` of those, one for each column, so ex. `set_icon(2, icon_texture)` becomes `columns[2].texture = icon_texture`
  - Make each of those properties `Array`s, with one slot for each column, so ex. `set_icon(2, icon_texture)` becomes `icons[2] = icon_texture`
- While we're at it, might as well rename `Tree`'s `columns` property to something like `column_count` to improve clarity to match the rest of the properties being renamed for `4.0`.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I think I've answered this one in a fair bit of detail above.  I'd need to familiarize myself more with how `Tree` is implemented in C++ to give any more specifics.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can be worked around with *way, way* too many lines of script (about 100 for my limited implementation of drag-and-drop alone, and that took me hours to figure out), including a lot of manually iterating through `TreeItem`s and some ugly hacks.

Also, it would be used only slightly less frequently than `Tree` itself.

### Is there a reason why this should be core and not an add-on in the asset library?

Improving core `Control` nodes (and giving them basic functionality) should not have to be done by add-ons.  If you're going to have the node at all, why not make it useful out of the box?

**Comment by Jummit** [comment]
Also, `edit_selected` could be exposed.

**Comment by name-here** [comment]
`PopupMenu`, illustrates one functional way to allow interaction with a similar tree-structured `Control` node, though I think I'd prefer arrays over a bunch of functions you pass an index/id to.  Especially nice (compared to `Tree`) is that you can edit a `PopupMenu` both through code _and_ through the editor!  Maybe `PopupMenu`'s graphical editor interface could be reused for `Tree`?

**Comment by MennoMax** [comment]
I totally agree with you that the tree control needs improvements, I worked on a modelling tool that uses cuboids (which I paused until the release of 4.0) and pretty much replicated the editors scene tree there which was rather painful. Here's my paused project, in case it helps you:

[CuboidWorks.zip](https://github.com/godotengine/godot-proposals/files/5866024/CuboidWorks.zip)

I'd really like to finish it in the future, so please only use it to learn from :)

**Comment by akien-mga** [comment]
Reopening, it was only *partially* addressed by https://github.com/godotengine/godot/pull/46773.

**Comment by me2beats** [comment]
> Getting the TreeNode at a particular index
we can make Tree more "Node-like" having methods add_child(), remove_child(), get_child().
Idk if it would be good in performance when having thousands of items tho

**Comment by me2beats** [comment]
> Adding an already-constructed TreeNode object to the Tree

maybe a new method like `create_from(TreeItem)`

**Comment by me2beats** [comment]
> Make TreeItem inherit from Node
Can't be done for performance reasons
the Tree would be slow when having thousands of items
See https://github.com/godotengine/godot-proposals/issues/4579#issuecomment-1135177259_

**Comment by YuriSizov** [comment]
Would be nice to re-evaluate if there is still something important missing. This is a very broad proposal, so it's hard to know if support comes for everything in it or for parts of it, and those parts may've been implemented already.

**Comment by dugramen** [comment]
> > Make TreeItem inherit from Node
> 
> Can't be done for performance reasons the Tree would be slow when having thousands of items See [#4579 (comment)](https://github.com/godotengine/godot-proposals/issues/4579#issuecomment-1135177259)_

Hi, I'm curious about this point. Can't any potential performance issues be mitigated by virtualization techniques. Like skipping collapsed items and items off screen. I imagine it wouldn't ever exceed a hundred nodes, and you'd only ever load/unload less than that amount at a time

**Comment by AnidemDex** [comment]
`Node` is still too bloated compared to what a `TreeItem` does. I don't think is a good idea to make TreeItem extends from node, but maybe that TreeItem (or tree?) imitates/exposes the same node tree that Node uses :thinking:

**Comment by dugramen** [comment]
> `Node` is still too bloated compared to what a `TreeItem` does. I don't think is a good idea to make TreeItem extends from node, but maybe that TreeItem (or tree?) imitates/exposes the same node tree that Node uses 🤔

Is it really? Looking through the docs, TreeItem also looks very bloated, while also being less flexible and more confusing. I feel like it would be cleaner to utilize existing control nodes, so users don't have to figure out this weird specific api instead of regular Godot (of course, only if performance could be acceptable)

On the flipside, a different api, which I find a lot more pleasant, is EditorProperty for inspector plugins. It has a bit of its own UI for inspector stuff, and then you're free to add whatever functionality you want via nodes.

**Comment by AnidemDex** [comment]
Sorry, I should have used different words, I was talking about the memory usage of those two objects compared, not about the amount of functions that they define
